### DISCUSSION TITLE
Knex raw select and sort

### DISCUSSION THREAD

**Discussion by HalikovStudioIT** [directus/directus#5840] [Archive]
Hello. How can i use knex.raw for select or sort in custom endpoints with use AST (to save check user permissions)?  
Maybe you have a simple solution for this?

**Comment by rijkvanzanten** [comment]
> to save check user permissions

You won't be able to use a raw SQL query _and_ keep Directus' permission system intact. That being said, you can use the `ItemsService` to fetch the items using any of the query params you'd normally find in the API, while keeping the permissions system. See https://docs.directus.io/guides/api-endpoints/#full-example for an example

**Reply by HalikovStudioIT** [reply]
I assumed so.
But what to do in a situation when, for example, you need to add sorting by a raw value (for example, by ST_Distance in PostGIS) or to select a raw value? Does this have to handle a completely raw request including user permissions and sub-requests (like getting a value from m2o)?
This could be worked around by adding raw values to exceptions if they are a Raw instance, but I understand that this may not be very useful and only used in rare cases.

**Reply by rijkvanzanten** [reply]
Sorting is already done through sql (order by) so that example doesn't really exist in this case. For every other database vendor specific operation, you'll have to resort to direct sql queries.

> This could be worked around by adding raw values to exceptions if they are a Raw instance, but I understand that this may not be very useful and only used in rare cases.

I don't know what you mean

**Reply by HalikovStudioIT** [reply]
I mean something like this:
```const service = new ItemsService('collection_name', { schema, accountability });
service.readByQuery({
  fields: [ '*', knex.raw('any raw') ],
  // or
  sort: [
    { column: knex.raw('any raw'), order: 'asc' }
  ]
});
```

**Reply by rijkvanzanten** [reply]
Ah I see. Yeah there's no plan to support that right now. Adds a lot of extra checks and overhead, for a very small flexibility gain. Don't think that's [80/20](https://docs.directus.io/contributing/introduction/#feature-requests) 🙂

**Reply by HalikovStudioIT** [reply]
I thought so, thanks :)
Is there an easy way to achieve this functionality? What I have just described looks monstrous and required additional export parseCurrentLevel, applyParentFilters and mergeWithParentItems from database/run-ast

**Reply by rijkvanzanten** [reply]
I'm not entirely sure what you're trying to do exactly.. Fields/sort already allows any column name, but it sounds like you need to run some custom SQL in that select?
